### ISSUE TITLE
Endpoint for getting all users online?

### ISSUE THREAD

**Issue by sunnydayoutside** [mk6i/retro-aim-server#6]
I'd like to create a frontend for my RAS server, but I think it would be really helpful if there was an endpoint that would report how much and who the people that are online.

**Comment by mk6i** [comment]
Sounds good to me. Are you interested in contributing this feature? If so, I can give you some direction on how to start. Otherwise, I can probably pick this up in a week or two.

**Comment by LushPineapple29** [comment]
> Sounds good to me. Are you interested in contributing this feature? If so, I can give you some direction on how to start. Otherwise, I can probably pick this up in a week or two.

I'm not familiar with Go, I'll try to implement it on my own if I can.

**Comment by mk6i** [comment]
Before you start, complete the [development environment guide](https://github.com/mk6i/retro-aim-server/blob/main/docs/BUILD.md).

Now let's say you want to add a `GET /user/online` endpoint that returns all users currently online. [InMemorySessionManager](https://github.com/mk6i/retro-aim-server/blob/main/state/session_manager.go) maintains the state of online users. All we need to do is have the endpoint dump the users returned by [InMemorySessionManager.AllSessions()](https://github.com/mk6i/retro-aim-server/blob/main/state/session_manager.go#L154).

Here's the gist of it:

1. Add the new route and HTTP handler to [server/http/mgmt_api.go](https://github.com/mk6i/retro-aim-server/blob/main/server/http/mgmt_api.go). Use the existing handlers as examples.
    ```diff
   + mux.HandleFunc("/user/online", uh.getOnlineUsers)
   //...
   + func (uh userHandler) getOnlineUsers(w http.ResponseWriter, r *http.Request) {
   +     // ...
   + }
    ```

2. Next, add the `InMemorySessionManager` dependency to user handler.
    ```diff
    type userHandler struct {
        UserManager
    +   state.InMemorySessionManager
        logger *slog.Logger
    }
    ```
  You'll need to make make the corresponding change to `StartManagementAPI()` and in `cmd/server/main.go`, where the dependencies are injected.

3. Finally, add logic to the new HTTP handler for pulling the list of users and sending it as a JSON response. (JSON response is not shown here).
    ```diff
    func (uh userHandler) getOnlineUsers(w http.ResponseWriter, r *http.Request) {
    // ...
    + onlineUsers := uh.InMemorySessionManager.AllSessions()
    + var screenNames []string
    + for _, sess := range onlineUsers {
    +     screenNames = append(screenNames, sess.ScreenName())
    + }
    // ...
    }
    ```

Let me know if/when you complete this, and then we can proceed to making the code tidy with interfaces, etc before the final commit. But this should guide you towards a working endpoint.

**Comment by LushPineapple29** [comment]
> Before you start, complete the [development environment guide](https://github.com/mk6i/retro-aim-server/blob/main/docs/BUILD.md).
> 
> Now let's say you want to add a `GET /user/online` endpoint that returns all users currently online. [InMemorySessionManager](https://github.com/mk6i/retro-aim-server/blob/main/state/session_manager.go) maintains the state of online users. All we need to do is have the endpoint dump the users returned by [InMemorySessionManager.AllSessions()](https://github.com/mk6i/retro-aim-server/blob/main/state/session_manager.go#L154).
> 
> Here's the gist of it:
> 
>     1. Add the new route and HTTP handler to [server/http/mgmt_api.go](https://github.com/mk6i/retro-aim-server/blob/main/server/http/mgmt_api.go). Use the existing handlers as examples.
>        ```diff
>        + mux.HandleFunc("/user/online", uh.getOnlineUsers)
>        //...
>        + func (uh userHandler) getOnlineUsers(w http.ResponseWriter, r *http.Request) {
>        +     // ...
>        + }
>        ```
> 
>     2. Next, add the `InMemorySessionManager` dependency to user handler.
>        ```diff
>        type userHandler struct {
>            UserManager
>        +   state.InMemorySessionManager
>            logger *slog.Logger
>        }
>        ```
> 
> 
> You'll need to make make the corresponding change to `StartManagementAPI()` and in `cmd/server/main.go`, where the dependencies are injected.
> 
>     3. Finally, add logic to the new HTTP handler for pulling the list of users and sending it as a JSON response. (JSON response is not shown here).
>        ```diff
>        func (uh userHandler) getOnlineUsers(w http.ResponseWriter, r *http.Request) {
>        // ...
>        + onlineUsers := uh.InMemorySessionManager.AllSessions()
>        + var screenNames []string
>        + for _, sess := range onlineUsers {
>        +     screenNames = append(screenNames, sess.ScreenName())
>        + }
>        // ...
>        }
>        ```
> 
> 
> Let me know if/when you complete this, and then we can proceed to making the code tidy with interfaces, etc before the final commit. But this should guide you towards a working endpoint.
> 
> Again, if this is too overwhelming, I'll take it on in the next couple weekends :)

Unfortunately, some stuff IRL just happened and I have to deal with that over this weekend. Sorry about that, I hope everything goes well and I hope to see this implemented soon!

**Comment by mk6i** [comment]
No problem. I'm reopening this ticket to track the work.

**Comment by mk6i** [comment]
Implemented in [v0.2.0](https://github.com/mk6i/retro-aim-server/releases/tag/v0.2.0). Please re-open this issue if you have questions or feedback.
